cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,302 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.notifications.service.kafka;
+
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.notifications.feeds.NotificationFeed;
+import co.cask.cdap.notifications.feeds.NotificationFeedException;
+import co.cask.cdap.notifications.feeds.NotificationFeedManager;
+import co.cask.cdap.notifications.service.NotificationException;
+import co.cask.cdap.notifications.service.NotificationHandler;
+import co.cask.cdap.notifications.service.NotificationService;
+import co.cask.cdap.notifications.service.inmemory.InMemoryNotificationService;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.common.Threads;
+import org.apache.twill.kafka.client.Compression;
+import org.apache.twill.kafka.client.FetchedMessage;
+import org.apache.twill.kafka.client.KafkaClient;
+import org.apache.twill.kafka.client.KafkaConsumer;
+import org.apache.twill.kafka.client.KafkaPublisher;
+import org.apache.twill.kafka.client.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Kafka implementation of the {@link NotificationService}.
+ */
+public class KafkaNotificationService extends AbstractIdleService implements NotificationService {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaNotificationService.class);
+  private static final Gson GSON = new Gson();
+
+  private static final int KAFKA_PUBLISHERS_CACHE_TIMEOUT = 3600;
+
+  private final KafkaClient kafkaClient;
+  private final NotificationFeedManager feedManager;
+  private final KafkaPublisher.Ack ack;
+  private final InMemoryNotificationService delegate;
+
+  // Cached instances of KafkaPublisher.Preparer to use to publish on different topics
+  private final LoadingCache<String, KafkaPublisher.Preparer> kafkaPublishersCache;
```

Comment:
But we do have one preparer per topic. Do you mean remove the cache in favor of a map?